x/exchange/types/matching.go

```diff
@@ -62,12 +63,21 @@ func (ctx *MatchingContext) fillOrder(orderType MemOrderType, isBuy bool, qty, p
 		pays = qty
 		receives = executedQuote
 	}
-	negativeMakerFeeRate := ctx.makerFeeRate.IsNegative()
-	if isMaker && negativeMakerFeeRate {
-		fee = ctx.makerFeeRate.MulTruncate(pays) // is negative
+	var feeRate sdk.Dec
+	if orderType == UserMemOrder {
+		feeRate = ctx.feeRate(isMaker)
+	} else {
+		if isMaker {
+			feeRate = ctx.takerFeeRate.Mul(ctx.orderSourceFeeRatio).Neg()
+		} else {
+			feeRate = utils.ZeroDec
+		}
+	}
+	if isMaker && feeRate.IsNegative() {
+		fee = feeRate.MulTruncate(pays)
 		pays = pays.Add(fee)
-	} else if orderType == UserMemOrder {
-		receives, fee = DeductFee(receives, ctx.feeRate(isMaker))
+	} else if !feeRate.IsZero() {
```

Comment:
If you check if it's not zero, the negative case also could be included, is that what you intended? If not, wouldn't it be right to check if it's positive?

Reply:
Nice suggestion, thanks.